src/harness/sas_objects.py

```diff
@@ -0,0 +1,353 @@
+     __init__
+          Parameters: grant ID, grant request
+    Retrieve  grant ID,
+    Retrieve  grant request
+    Retrieve operation params
+    Construct heartbeat request
+    Update was authorized in last heartbeat? state
+    Retrieve was authorized in last heartbeat
+    """
+
+    def __init__(self, g_id, g_request):
+        self.grant_id = g_id
+        self.grant_request = g_request
+        self.grant_authorized_in_last_heartbeat = False
+
+    def get_grant_id(self):
+        """
+
+        :rtype: grant_id
+        """
+        return self.grant_id
+
+    def get_grant_request(self):
+        """
+
+        :rtype: grant_request
+        """
+        return self.grant_request
+
+    def get_request_operation_param(self):
+        """
+
+        :rtype: operationParam
+        """
+        return self.grant_request['grantRequest'][0]['operationParam']
+
+    def construct_heartbeat_request(self):
+        """
+
+        :rtype: heartbeat_request
+        """
+        heartbeat_request = {
+            'heartbeatRequest': [{
+                'grantId': self.grant_id,
+                'cbsdId': self.grant_request['grantRequest'][0]['cbsdId'],
+                'operationState': "GRANTED"
+            }]
+        }
+        return heartbeat_request
+
+    def update_grant_authorized_in_last_heartbeat(self, is_authorized):
+        self.grant_authorized_in_last_heartbeat = is_authorized
+
+    def get_grant_authorized_in_last_heartbeat(self):
+        """
+
+        :rtype: grant_authorized_in_last_heartbeat
+        """
+        return self.grant_authorized_in_last_heartbeat
+
+
+class Cbsd:
+
+    """
+    A CBSD object has the following methods
+        __init__
+            Parameters: CBSD ID, successful registration request, grant ID, successful grant request
+        Retrieve   CBSD ID,
+        Retrieve successful registration request,
+        Retrieve grant request, successful
+            Parameter: grant request
+        Does this CBSD have at least one active grant?
+        Construct heartbeat request for all active grants
+        Return operating params for all grants which are authorized to transmit
+    """
+    def __init__(self, c_id, reg_request, g_id, grant_requests):
+        self.cbsd_id = c_id
+        self.grant_objects = {}
+        for index, grantRequest in enumerate(grant_requests):
+            self.grant_objects[g_id[index]] = Grant(g_id[index], grantRequest)
+        self.successful_registration_request = reg_request
+
+    def get_cbsd_id(self):
+        """
+
+        :rtype: cbsd_id
+        """
+        return self.cbsd_id
+
+    def get_grant_request(self, grant_id):
+        """
+
+        :rtype: grant_objects
+        """
+        return self.grant_objects[grant_id]
+
+    def get_registration_request(self):
+        """
+
+        :rtype: successful_registration_request
+        """
+        return self.successful_registration_request
+
+    def cbsd_id_has_atleast_one_grant(self):
```

Comment:
"at least" is two words. A CBSD has a Grant, not a CBSD ID. 

Based on the body of the function, this is retrieving only authorized Grants. If that is the intention, a more appropriate name would be `cbsd_has_at_least_one_authorized_grant`.

However, I think we need a function that checks to see whether or not the CBSD has any *active* (i.e. non-terminated/non-relinquished) Grants. This will allow us to quickly pass over any CBSDs which no longer have Grants in some of our other functions.

Reply:
Fixed as   per suggestion